www/nginx/src/opnsense/scripts/nginx/ngx_autoblock.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+/*
+ * Copyright (C) 2018 Fabian Franz
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+require_once('config.inc');
+require_once('IPv6.inc');
+
+use OPNsense\Firewall\Alias;
+use OPNsense\Nginx\AccessLogParser;
+use OPNsense\Core\Config;
+
+function nginx_print_error($msg) {
+    echo json_encode(
+        array('status' => 'error', 'message' => $msg)
+    );
+}
+function exec_hidden($command): void
+{
+    $descriptorspec = array(
+        1 => array('file', "/dev/null", 'w'),
+        2 => array('file', "/dev/null", "w")
+    );
+
+    $process = proc_open($command, $descriptorspec, $pipes);
+
+    if (is_resource($process)) {
+        proc_close($process);
+    }
+}
+function add_to_blocklist($tablename, $ip) {
+    $escaped = escapeshellarg($ip);
+    echo $ip . PHP_EOL;
+    exec_hidden("/sbin/pfctl -t ${tablename} -T add ${escaped}");
+}
+
+
+function reopen_logs() {
+    exec_hidden('/usr/local/sbin/nginx -s reopen');
+}
+
+$permanent_ban_file = '/var/log/nginx/permanentban.access.log';
+$permanent_ban_file_work = $permanent_ban_file . '.work';
+$autoblock_alias_name = 'nginx_autoblock';
+
+
+if (!file_exists($permanent_ban_file)) {
+    nginx_print_error('No Log exists - nothing to do');
+    // let create it
+    reopen_logs();
+    exit(0);
+}
+
+// move the file, and inform nginx that we deleted the file
+rename($permanent_ban_file, $permanent_ban_file_work);
+reopen_logs();
+
+$log_parser = new AccessLogParser($permanent_ban_file_work);
+
+$log_lines = $log_parser->get_result();
+
+$model = new Alias();
+
+$blacklist_element = null;
+foreach ($model->aliases->alias->__items as $alias) {
+    if ((string)$alias->name == $autoblock_alias_name) {
+        if ((string)$alias->type != 'network') {
```

Comment:
@AdSchellevis can you explain the external field? It seems to have no validations and the current one works. I want the user to be able to delete an entry if it detected an incorrect one.

The script runs once per minute and appends new findings to the alias and updates the alias by itself when done. The log is purged automatically but the IP addresses keep being stored in the alias.

Reply:
@fabianfrz external means that it’s managed elsewhere (in this case in your plugin), the alias system won’t touch the contents and trusts the provider to manage it accordingly. With the current construction you created two systems, which try to update the table. The pfct in your code and the regular alias system updating the contents to its configuration at given intervals, which causes a race and can lead to removal of entries just inserted by your tool. 

For better separation of concerns it’s highly advisable to keep the blacklist in your plugin or ask the alias system to sync it’s contents, my advise is to keep the blacklist in the plugin since it causes less overhead. People can always drop entries manually using the table tool, same as for ssh an web ui lockout construction.

Reply:
so external will do nothing except adding a table entry to pf.conf?

Reply:
Yes, that’s it